flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/OnlineKMeans.java

```diff
@@ -0,0 +1,407 @@
+     *   <li>Computes the new centroids from the average of input points that belongs to the same
+     *       cluster
+     *   <li>Computes the weighted average of current and new centroids. The weight of a new
+     *       centroid is the number of input points that belong to this cluster. The weight of a
+     *       current centroid is its original weight scaled by $ decayFactor / parallelism $.
+     *   <li>Generates new model data from the weighted average of centroids, and the sum of
+     *       weights.
+     * </ul>
+     */
+    private static class ModelDataLocalUpdater extends AbstractStreamOperator<KMeansModelData>
+            implements TwoInputStreamOperator<DenseVector[], KMeansModelData, KMeansModelData> {
+        private final DistanceMeasure distanceMeasure;
+        private final int k;
+        private final double decayFactor;
+        private ListState<DenseVector[]> localBatchDataState;
+        private ListState<KMeansModelData> modelDataState;
+
+        private ModelDataLocalUpdater(DistanceMeasure distanceMeasure, int k, double decayFactor) {
+            this.distanceMeasure = distanceMeasure;
+            this.k = k;
+            this.decayFactor = decayFactor;
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+
+            TypeInformation<DenseVector[]> type =
+                    ObjectArrayTypeInfo.getInfoFor(DenseVectorTypeInfo.INSTANCE);
+            localBatchDataState =
+                    context.getOperatorStateStore()
+                            .getListState(new ListStateDescriptor<>("localBatch", type));
+
+            modelDataState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>("modelData", KMeansModelData.class));
+        }
+
+        @Override
+        public void processElement1(StreamRecord<DenseVector[]> pointsRecord) throws Exception {
+            localBatchDataState.add(pointsRecord.getValue());
+            alignAndComputeModelData();
+        }
+
+        @Override
+        public void processElement2(StreamRecord<KMeansModelData> modelDataRecord)
+                throws Exception {
+            Preconditions.checkArgument(modelDataRecord.getValue().centroids.length == k);
+            modelDataState.add(modelDataRecord.getValue());
+            alignAndComputeModelData();
+        }
+
+        private void alignAndComputeModelData() throws Exception {
+            if (!modelDataState.get().iterator().hasNext()
+                    || !localBatchDataState.get().iterator().hasNext()) {
+                return;
+            }
+
+            KMeansModelData modelData =
+                    OperatorStateUtils.getUniqueElement(modelDataState, "modelData")
+                            .orElseThrow((Supplier<Exception>) NullPointerException::new);
```

Comment:
Given that `modelDataState.get().iterator().hasNext() == true`, the `orElseThrow()` should not be triggered, right?

Would it be simpler to just call `get()` here?

Reply:
`hasNext() == true` guarantees that there is 1 or more elements in the state, but `getUniqueElement` requires that there should be only 1 element in the state, so the `orElseThrow` would still be triggered if there are more than 1 elements.

With that said, I think the `NullPointerException` might be improper here. I'll change it to `IllegalStateException`.

Reply:
`getUniqueElement()` also guarantee that there should be exactly 1 element. This is because it throws `IllegalStateException` if there are more than 1 element in the state.

Reply:
Got it. As `getUniqueElement` has already thrown the exception, I'll change the implementation here to `get()`.